### PR TITLE
Helper functions and examples for modifier volumes and Z clipping

### PR DESCRIPTION
**Pull request by Twada2004**
It's nice to meet all the KOS developers.
After seeing the recent amazing announcements of 3D games, I wanted to do something to help.
So I created a sample modifier volume, which I hope will be helpful to KOS users.
This is my first pull request, so there are many things I don't understand. Please be gentle.

IMG_2051

Modifier volumes have the following constraints.
1. They must be closed polygons.
2. There must be a header indicating this just before the first and last polygons.

The requirement that they must be closed polygons is a big problem when it comes to Z-clipping.
It seems that modifier volumes are determined to be inside if the number of polygons blocking the camera is odd, and outside if the number is even.
Therefore, when the camera enters the volume, a covering polygon must be generated.
In this sample, if Z-clipping occurs, a covering polygon is generated in the shape of a triangle fan.
The covering polygons of a triangle fan are always odd inside and always even outside. It doesn't matter if there are cavities or enclaves inside the volume.
This algorithm is based on a Sega patent (2000-163601) published in Japan on June 16, 2000. The patent has been withdrawn and it is over 20 years old, so it can be used freely.

Because a header is required just before the first and last polygon, this sample helper function sends it to TA for each volume.
The first header is sent by applying a bit mask to the last header, so only the last header is passed to the function.
I found a bug in pvr_mod_compile. A bit needs to be set for the last header.
pvr_mod_compile(&hdr, PVR_LIST_OP_MOD, PVR_MODIFIER_INCLUDE_LAST_POLY, PVR_CULLING_SMALL);
hdr.cmd |= (1 << 6); /* Last poly */

Coordinates are received as (x, y, 1/z) and clipped to the range (1.0f >=z&&z>0.0f).
Use double-sided polygons.
Uses pair single precision transfer commands, so data should be 8-byte aligned.
I hope someone will improve this.

### CODE REVIEW

**Comment by gyrovorbis** [comment]
@Twada2004 YEEEEEEEEEEEEEEEEEEEEEEEEES!!!!!!!!!!!!!!!!!!!!!!!!!! SO GOOD TO SEE YOU!!!! Will be taking a closer look here shortly! I'm still testing your Bloom/HDR demo with the tip of KOS, btw. lmao.

File: examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h:1]: I still need to review these, but wow... These routines just seem too useful and generic to just be a utility file in only this example... I am of the opinion this should be added to KOS itself in like `kos/kernel/arch/dreamcast/include/dc/pvr/pvr_zclip.h` and then probably include this file within the overall `pvr.h` in the containing directory.

What do you think? Do you think these routines are generalized enough that they should be part of KOS itself? As a sub API within the PVR API? This will let anyone using KOS use them, plus they can be reused with other examples as well....

Comment (gyrovorbis) [reply examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h:1]: @Kazade what do you think? Have a look at these z clipping routines...

File: examples/dreamcast/pvr/modifier_volume_zclip/Makefile
Comment (pcercuei) [inline examples/dreamcast/pvr/modifier_volume_zclip/Makefile:29]: Instead of defining your own rules for the romdisk, just define `KOS_ROMDISK_DIR`, and the correct rules will be used automatically.

**Comment by Twada2004** [comment]
@gyrovorbis
I couldn't be happier if my samples are useful!
The Dreamcast still has a lot of potential. Thank you for your work!

@pcercuei
Thanks for reviewing the code!
I noticed for the first time that romdisk is easier to use.
I changed the romdisk code based on other samples. What do you think?

**Comment by gyrovorbis** [comment]
Just wanted to let you know, I've actually added the code to KOS itself to align `matrix_t` types to 8-bytes and just opened a PR here: https://github.com/KallistiOS/KallistiOS/pull/678 to align all PVR vertex and header types to 32-bytes... I think it's best to just add that to KOS, as it's less work on the user and is less error-prone. Once this PR which adds the alignment gets merged into master, you can merge this PR back onto master and remove all of the manual `__attribute__((aligned()))` clauses from your PR. :)

**Comment by pcercuei** [comment]
> @pcercuei Thanks for reviewing the code! I noticed for the first time that romdisk is easier to use. I changed the romdisk code based on other samples. What do you think?

Looks good!

**Comment by gyrovorbis** [comment]
@Twada2004 Sorry! I forgot to mention that I had to leave for a week and a half for a business trip for my job... Just got back home and plan to continue reviewing this.

One thing I'm unsure of here is your `pvr_zclip.h/.c` files. Are these routines generalized enough that they can be used by others for other applications, or are the parameters hardcoded for this specific example?

I feel like those routines and files are so useful that they should probably go directly into KOS where they can get used by anyone using KOS, rather than just having them in this one example... What do you think? Are the routines generic enough to be part of KOS lib?

**Comment by Twada2004** [comment]
@gyrovorbis Please take plenty of rest.

I'd be very happy if you could add this to KOS.
I created pvr_zclip to be as generic as possible so that you can copy the file and use it immediately.
This function is designed to receive coordinates with Z division already done, output by mat_trans_single and mat_trans_single4.
The hard-coded value is that the clipping plane is Z=1. It uses the projection matrix, multiplies the near Z value, and then divides by Z.

You should be able to specify any clipping plane by modifying it as follows.
1. Set the interpolation coefficient 1.0f to any value.
float inter = (inside->z - near_z) / (inside->z - outside->z);
2. Set the Z comparison condition to any value.
if (near_z >= src->z && src->z > 0.0f)

Since the PVR coordinate system takes a value of 1/Z, I think Z=1 is the most general value, but if necessary, you can modify it to receive any Z value.


The other part that is hard-coded is sending vertices directly to TA using the store queue.
It should work the same as pvr_prim. Therefore, since it does not support pvr_list_prim, you must modify it or create a special command if you want to use DMA.
I don't really understand the difference between the recently added direct rendering function and pvr_prim...
The function I use privately allows you to specify a separate output destination to support intermediate buffers.
int pvr_vertex_commit_zclip(pvr_vertex_t *dest, pvr_vertex_t *src, int size)
Like the bloom demo. It's not simple, so I thought it might be hard to use. Is this better?

**Comment by gyrovorbis** [comment]
Just wanted to let you know that I just got this PR through: https://github.com/KallistiOS/KallistiOS/pull/678.

I saw that anyone who wanted good performance was needing to align their vertices and polygon headers, so now that's done automatically. I also took care of it for the KOS `matrix_t` type as well, awhile ago, so we should remove the manual alignment directives for those types from the code.

**Comment by Twada2004** [comment]
> Just wanted to let you know that I just got this PR through: #678.
> 
> I saw that anyone who wanted good performance was needing to align their vertices and polygon headers, so now that's done automatically. I also took care of it for the KOS type as well, awhile ago, so we should remove the manual alignment directives for those types from the code.`matrix_t`

Updated to support the latest KOS!

File: examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h
Comment (QuzarDC) [reply examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h:1]: I'd agree generally with this, based on what I understand of how they'd be used. That should come with significant documentation in the header and (where possible) a reduction or separating out of the inlined assembly (generally in the codebase we try to avoid it except when needed or when it provides a substantial difference in efficiency). 

In either case though, I don't think that should need to hold up the inclusion of the examples. If you'd want to take up including that into the core pvr, then an issue could be made to track that as future work and in the meantime people could have access to this.

Comment (gyrovorbis) [reply examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h:1]: Oh... well, shit... OKAY. Maybe this is the path forward... I did really want this inside of KOS proper, but man, that's asking a lot of him for a first-time contribution which could already be benefiting the community... Huh...

So yeah, maybe we should take this as-is, then I can make a ticket for myself to do the refactor once I'm ready to add some of this other stuff GTA3 is bringing to the table in these areas too... do it all at once....

File: examples/dreamcast/pvr/modifier_volume_zclip/Makefile
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/Makefile:1]: #
# Modifier Volume Z-Clipping Example
# Copyright (C) 2024 Twada
#

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:1]:
```suggestion
/*
 * KallistiOS ##version##
 *
 * examples/dreamcast/pvr/modifier_volume_zclip/example.c
 * Copyright (C) 2024 Twada
 *
 * This example demonstrates how to perform Z-clipping on modifier volumes.
 */
```

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:10]:
```suggestion
static pvr_ptr_t box_tex;
```

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:12]:
```suggestion
static void mul_screen(float width, float height)
```

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:26]:
```suggestion
static void mul_projection(float fov, float aspect, float znear)
```

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:43]:
```suggestion
static void draw_modifier(float *pvm)
```

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:119]:
```suggestion
static void draw_box(float *pvm)
```

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:196]:
```suggestion
static void draw_plane(float *pvm)
```

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/example.c:200]:
```suggestion
    static pvr_vertex_t poly[4] = {
```
I just actually got these aligned to 32-bytes in the master branch for 2.1.0 release.

File: examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.c
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.c:1]:
```suggestion
/*
 * KallistiOS ##version##
 *
 * examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.c
 * Copyright (C) 2024 Twada
 *
 * This file contains the implementations of commonly-used Z-clipping routines for
 * modifier volumes. 
 */

#include "pvr_zclip.h"
```

File: examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h
Comment (gyrovorbis) [inline examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h:1]:
```suggestion
/*
 * KallistiOS ##version##
 *
 * examples/dreamcast/pvr/modifier_volume_zclip/pvr_zclip.h
 * Copyright (C) 2024 Twada
 *
 * This file contains the public API for commonly-used Z-clipping routines for
 * modifier volumes. 
 */

#include <kos.h>
```

**Review by gyrovorbis — COMMENTED** [review]
OKAY, MAN!!!

SO SO SO SORRY for getting back to you so late. I wasn't really supposed to be merging anything into `master` other than bugfixes or documentation until after the `2.1.0` release, which is now done! So one of the very first things I wanted to finish up was this fantastic PR.

Based on what @QuzarDC said, *I* can personally do the work to move the Z clipping routines into KOS itself in a follow-up PR... So if you can address the changes I advised to the code (just stylistic things), I can go ahead and merge this for everyone to start using and learning from it!

File: examples/dreamcast/pvr/modifier_volume_zclip/example.c
Comment (gyrovorbis) [reply examples/dreamcast/pvr/modifier_volume_zclip/example.c:10]: We make all globals which are only used within the C file they are declared within as being `static`.

**Comment by Twada2004** [comment]
> よし、お母さん!!
> 
> とても遅くなって戻ってきてすみません。リリース後まで、バグ修正やドキュメント以外のものにマージするはずではなかったのですが、今では完了しています!ですから、最初に仕上げたかったことの1つは、この素晴らしいPRでした。`master``2.1.0`
> 
> そうは言ってもらったことに基づいて、_私は_個人的にフォローアップPRでZクリッピングルーチンをKOS自体に移動する作業を行うことができます...したがって、私がコードにアドバイスした変更(スタイル上のもの)に対処できる場合は、先に進んでこれをマージして、誰もが使用して学び始めることができます。
> 
> 編集:ところで、将来的には、このようなPRがはるかに迅速にマージされるようにするためのはるかに優れたプロセスがあります...ただお知らせしたかったのですが、このような高品質な作品が届くまでに時間がかかったことを残念に思います...続けてください、私たちはそれを愛してきました、そしてこれは非常に多くの人に利益をもたらすでしょう!



> OKAY, MAN!!!
> 
> SO SO SO SORRY for getting back to you so late. I wasn't really supposed to be merging anything into `master` other than bugfixes or documentation until after the `2.1.0` release, which is now done! So one of the very first things I wanted to finish up was this fantastic PR.
> 
> Based on what @QuzarDC said, _I_ can personally do the work to move the Z clipping routines into KOS itself in a follow-up PR... So if you can address the changes I advised to the code (just stylistic things), I can go ahead and merge this for everyone to start using and learning from it!
> 
> EDIT: By the way, in the future we have a much better process in place for ensuring PRs like this get merged in much more quickly... Just wanted to let you know, as I feel bad that such high quality work has taken so long to make it in... Please do continue, we have loved it, and this will benefit so many!

Thank you for the code review!
And congratulations on the release of KOS2.1.0!

**Review by gyrovorbis — CHANGES_REQUESTED** [review]

**Review by gyrovorbis — APPROVED** [review]
OMG I APPROVE!!!! WHOOO!!!!

**Comment by gyrovorbis** [comment]
Congratulations, my friend!!! I promise it will never be that painful again! :rofl:

**Comment by Twada2004** [comment]
I'm thrilled to be able to contribute to KOS! Thank you so much for all your reviews.

**Pull request merged**
